rescript/get_midori2.py

```diff
@@ -0,0 +1,103 @@
+from urllib.error import HTTPError
+from pathlib import Path
+import pandas as pd
+from q2_types.feature_data import (TaxonomyFormat, DNAFASTAFormat,
+                                   DNAIterator)
+
+
+def _assemble_midori2_urls(mito_gene,
+                           version='GenBank265_2025-03-08',
+                           ref_seq_type='uniq',
+                           unspecified_species=False,
+                           ):
+    print('Building URLs ...')
+    if unspecified_species:
+        spec = 'QIIME_sp'
+        sp = 'SP_'
+    else:
+        spec = 'QIIME'
+        sp = ''
+
+    vernum = version.strip('GenBank').split('_')[0]
+
+    base_url = ('https://www.reference-midori.info/download/Databases/'
+                '{full_version}/{spec}/{ref_seq_type}/MIDORI2_'
+                '{ref_seq_type_u}_NUC_'
+                '{sp}GB{vernum}_{mito_gene}_QIIME.{file_type}.gz')
+
+    base_format_dict = {'full_version': version,
+                        'spec': spec,
+                        'ref_seq_type': ref_seq_type,
+                        'ref_seq_type_u': ref_seq_type.upper(),
+                        'sp': sp,
+                        'vernum': vernum,
+                        'mito_gene': mito_gene}
+
+    fasta_url = base_url.format(**base_format_dict, **{'file_type': 'fasta'})
+    tax_url = base_url.format(**base_format_dict, **{'file_type': 'taxon'})
+
+    return fasta_url, tax_url
+
+
+def _retrieve_data_from_midori2(fasta_url, tax_url):
+
+    print('\nDownloading and processing raw files ... \n')
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        for url in [fasta_url, tax_url]:
+            compressed_fn = Path(url).name
+            uncompressed_fn = Path(url).stem
+            in_path = os.path.join(tmpdirname, compressed_fn)
+            out_path = os.path.join(tmpdirname, uncompressed_fn)
+
+            try:
+                print('Retrieving data from {0}'.format(url))
+                urlretrieve(url, in_path)
+            except HTTPError:
+                raise ValueError('Unable to retrieve the following file '
+                                 'from MIDORI2:\n' + url)
+
+            print('  Unzipping {0}...\n'.format(in_path))
+            with gzip.open(in_path, 'rt') as gz_in:
+                with open(out_path, 'w') as gz_out:
+                    shutil.copyfileobj(gz_in, gz_out)
+                    if out_path.endswith('fasta'):
+                        seqs = DNAFASTAFormat(out_path,
+                                              mode="r").view(DNAIterator)
+                    elif out_path.endswith('taxon'):
+                        tax = TaxonomyFormat(out_path,
+                                             mode="r").view(pd.DataFrame)
+    return seqs, tax
+
+
+def get_midori2_data(
+    mito_gene: str,
+    version: str = 'GenBank265_2025-03-08',
+    ref_seq_type: str = 'uniq',
+    unspecified_species: bool = False,
+        ) -> (DNAIterator, pd.DataFrame):
```

Comment:
@VinzentRisch, assuming that I update given your other comments... I should set `mito_gene: list`. But what do I set the expected output to?  I'd imagine it can't still be `-> (DNAIterator, pd.DataFrame)`.

Reply:
Got it working! 🎆